doc/dev/kdb-record.md

```diff
@@ -0,0 +1,110 @@
+# `kdb record` Command
+
+The `kdb record` tooling enables the user to record configuration sessions. It also allows exporting and importing the configuration changes.
+
+Recording sessions are stored on a per-user basis. This implies that a recording session can only track changes made by a single user.
+
+## Session handling
+
+* `kdb record start [session_name]`: Starts a new recording session. If no session name is provided, a new random name is generated. If a session with the provided name already exists, changes will be appended.
+* `kdb record stop`: Stops the current recording session. 
+* `kdb record delete [session_name]`: Deletes the specified recording session. If no session name is provided, the current session will be deleted. Deleting a session will not stop the recording. 
+* `kdb record list-sessions`: Lists all sessions by their name
+* `kdb record changes [session_name]`: Shows all changes for the specified session. If no session name is provided, the current session will be used. 
+
+## Exporting configuration changes
+
+`kdb record export [session_name] [--format <format>]`: Will dump the changed keys of the specified session onto `stdout`. If no session name is provided, the current session will be used. 
+
+As example, suppose the key `user:/my/app/hello` has the value `world` and you execute the following commands:
+* `kdb set user:/my/app/hello max`
+* `kdb set user:/my/app/hello susi`
+
+Only the value `susi` will be exported. Additionally, if you execute `kdb set user:/my/app/hello world` the key will not be exported at all, as there are no changes compared to the beginning of the recording.
```

Comment:
> "replay" is probably export and then import with merge.

I think that's unnecessarily complex. Directly applying all the recorded changes one-by-one and then doing `kdbSet` at the end (to make the abort in case of conflicts easy), seems much simpler.

Reply:
Just to be clear: should `replay` export something so that it can be used on another machine/kdb instance, or is it only a "redo" if you "undo" the changes of this session?

Reply:
I was thinking about `replay` applying the changes immediately, but writing to a file could be even better. But my main point was that `kdb record export` seems to produce the diff between start and end. So something like this

```sh
kdb set user:/foo/bar abc

kdb record start
kdb set user:/foo/bar 123
kdb set user:/foo/bar abc
kdb record export
```

produces an empty export file, since `user:/foo/bar` had the same value at the start and at the end.

With `kdb record replay` the intermediate change should still happen, because it could have side effects (depending on what plugins do).

Reply:
> Just to be clear: should replay export something so that it can be used on another machine/kdb instance, or is it only a "redo" if you "undo" the changes of this session?

Yes, we need to distinguish these use cases clearly. The minimum local functionality is `undo` (`redo`, `replay` is probably not needed so often). The more interesting part is the export+import functionality, which must continue to work after modifications, ergo 3-way merge yields better results.

Reply:
IMO this really depends on what `export` actually does. If it ignores intermediate changes, some other command that doesn't is also essential. It could also be an option for the `kdb record export` command. But if we cannot export the full log of changes, the tool is just a diff tool and not a proper recording tool.